### PR TITLE
Reduce boilerplate for BytePos and CharPos

### PR DESCRIPTION
**Pull request by Juici**
Reduces boilerplate code for BytePos and CharPos by using a macro to implement shared traits.

### CODE REVIEW

File: compiler/rustc_span/src/lib.rs [hunk 1]
```diff
@@ -1558,58 +1558,71 @@ pub trait Pos {
...
+                type Output = $ident;
 
-    #[inline(always)]
-    fn to_u32(&self) -> u32 {
-        self.0
-    }
-}
+                #[inline(always)]
+                fn add(self, rhs: $ident) -> $ident {
+                    $ident((self.to_usize() + rhs.to_usize()) as $inner_ty)
```
Comment (lcnr) [inline compiler/rustc_span/src/lib.rs:1599]:
```suggestion
                    $ident(self.0 + rhs.0)
```
I don't think there is anything requiring us to go through `usize` here, is there?

**Review by lcnr — APPROVED** [review]
Thanks :+1: 

I have 1 small nit but otherwise this already LGTM.

File: compiler/rustc_span/src/lib.rs (hunk 1 above)
Comment (Juici) [reply compiler/rustc_span/src/lib.rs:1599]: I don't believe so, but I assumed that there was a reason the cast in the old code.

Comment (lcnr) [reply compiler/rustc_span/src/lib.rs:1599]: I would expect there to be a comment if this is the case, so I would prefer changing this.

Comment (Juici) [reply compiler/rustc_span/src/lib.rs:1599]: That makes sense, it'd be better to error instead of transparently truncating.

**Comment by lcnr** [comment]
Thanks

@bors r+ rollup

**Pull request merged**